### DISCUSSION TITLE
Fehler bei Update: "1091 Can't DROP FOREIGN KEY `notes?note_type_id?note_types?id`"

### DISCUSSION THREAD

**Discussion by habreli** [tine20/tine20#7387] [General]
Die gesamte Fehlermeldung lautet:

`SQLSTATE[42000]: Syntax error or access violation: 1091 Can't DROP FOREIGN KEY `notes?note_type_id?note_types?id`; check that it exists, query was: ALTER TABLE `tine20_notes` DROP FOREIGN KEY `notes??note_type_id??note_types??id``

Ich versuche gerade ein Update von der Version "2022.06.04.96" auf die letzte Version bzw. auch auf die Version "2022.06.25.100" - beide verhalten sich gleich.

Ich habe/hatte bei dieser Tabelle den genannten ForeignKey nicht.
Bei mir gibt’s auch die Felder "notes" und "note_types" nicht.

Wie komme ich nun weiter?
Hat das schon jemand gelöst?

Danke im Voraus.

**Comment by ignoreigor** [comment]
Guten Morgen,

einen FK habe ich auch nicht (mehr), wohl aber einen Key

| tine20_notes | CREATE TABLE `tine20_notes` (
  `id` varchar(40) COLLATE utf8mb4_unicode_ci NOT NULL,
  `note_type_id` varchar(40) COLLATE utf8mb4_unicode_ci NOT NULL,
  `note` mediumtext COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `record_model` varchar(64) COLLATE utf8mb4_unicode_ci NOT NULL,
  `record_backend` varchar(40) COLLATE utf8mb4_unicode_ci NOT NULL,
  `record_id` varchar(40) COLLATE utf8mb4_unicode_ci NOT NULL,
  `created_by` varchar(40) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `creation_time` datetime DEFAULT NULL,
  `last_modified_by` varchar(40) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `last_modified_time` datetime DEFAULT NULL,
  `is_deleted` tinyint(1) unsigned NOT NULL DEFAULT 0,
  `deleted_by` varchar(40) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `deleted_time` datetime DEFAULT NULL,
  `seq` int(11) unsigned NOT NULL DEFAULT 0,
  PRIMARY KEY (`id`),
  KEY `record_id` (`record_id`),
  KEY `record_model` (`record_model`),
  KEY `record_backend` (`record_backend`),
  **KEY `notes::note_type_id--note_types::id` (`note_type_id`)**
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci |

Eine Tabelle tine20_notes_types gibt es wohl nicht mehr. Von daher brauche ich mich nicht zu fragen, ob das wirklich als Fremdschlüssel funktionieren würde.

Bin mir nicht sicher, ob ich die Fehlermeldung auch mal gesehen hatte und dachte, na wenn er nicht da ist, braucht er ihn nicht zu löschen...

**Comment by ignoreigor** [comment]
Im Backup vom  2022-07-04 habe ich noch eine FK-Contraint und auch die dazugehörige Tabelle, am 23.07. schaut es so aus wie jetzt und die Tabelle fehlt.

:CREATE TABLE `tine20_notes` (
  `id` varchar(40) COLLATE utf8mb4_unicode_ci NOT NULL,
  `note_type_id` varchar(40) COLLATE utf8mb4_unicode_ci NOT NULL,
  `note` mediumtext COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `record_model` varchar(64) COLLATE utf8mb4_unicode_ci NOT NULL,
  `record_backend` varchar(40) COLLATE utf8mb4_unicode_ci NOT NULL,
  `record_id` varchar(40) COLLATE utf8mb4_unicode_ci NOT NULL,
  `created_by` varchar(40) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `creation_time` datetime DEFAULT NULL,
  `last_modified_by` varchar(40) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `last_modified_time` datetime DEFAULT NULL,
  `is_deleted` tinyint(1) unsigned NOT NULL DEFAULT 0,
  `deleted_by` varchar(40) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `deleted_time` datetime DEFAULT NULL,
  `seq` int(11) unsigned NOT NULL DEFAULT 0,
  PRIMARY KEY (`id`),
  KEY `record_id` (`record_id`),
  KEY `record_model` (`record_model`),
  KEY `record_backend` (`record_backend`),
  KEY `notes::note_type_id--note_types::id` (`note_type_id`),
  **CONSTRAINT `notes::note_type_id--note_types::id` FOREIGN KEY (`note_type_id`) REFERENCES `tine20_note_types` (`id`)**
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;

Vermute es kommt aus Funktion update013 in Datei ./Tinebase/Setup/Update/15.php  und es hätte die Tinebase-Version auf 15.13 hochziehen müssen, kannst ja mal schauen, welche Version die Tinebase-App trägt. Mittlerweile (jüngstes Release) trägt sie bei mir 15.15.

**Reply by habreli** [reply]
Tja - irgendwie triffst du ins Schwarze. Auch wenns ein Bug oder so sein dürfte ? (ich weiß - böses Wort)
Die Tabelle "tine20_notes_types" hab ich noch.
Und Tinebase-App ist bei mir 15.12.
Den CONSTRAINT Key gabs bei bei mir nicht. Und den letzten Key hab ich manuell gelöscht (bringt aber eh nichts).
Eigentlich sollte die Update Routine doch weiter laufen (so ähnlich Try & Catch) wenn der Schlüssel nicht da ist.
Ich kann die Zeile wohl auskommentieren - aber bei der nächsten Version häng ich ja wieder....

**Comment by ignoreigor** [comment]
Die Update-Routine wurde tatsächlich in Release 2022.06.25.100 eingeführt.

Lustigerweise habe ich gerade mein altes Issue #7385 gefunden, bei dem mich just diese FK-Constraint in dieser Version daran hinderte, Termine zu aktualisieren. Das deutet darauf hin, dass ich trotz des erfolgten Updates auf 2022.06.25.100 noch die FK aktiv hatte. Also umgekehrt, wie bei Dir. 

Du könntest die  beiden Befehle dropForeignKey und dropTable auskommentieren und die Updates der ID machen. Danach sollte ja deine Tinebase-Version aktualisiert werden. In dem kommenden Versionen würde dieses Update dann nicht mehr durchlaufen, von daher ist es auch nicht tragisch, wenn die Auskommentierung wieder weg wäre.

**Reply by habreli** [reply]
Danke für die Unterstützung,
so hab ichs auch tatsächlich gelöst bekommen.
